README.md
# ComicBook

漫画爬虫、漫画下载工具，支持腾讯漫画、哔哩哔哩漫画、有妖气漫画、快看漫画、漫画柜等站点

[漫画源收录情况](https://github.com/lossme/ComicBook/projects/1)

尊重版权，请支持正版，通过本工具下载或生成的资源**禁止传播分享！禁止利用本项目进行商业活动！**

## 本项目特点

- [x] 漫画批量下载
- [x] 分目录按章节保存
- [x] 支持多个漫画源，易于扩展
- [x] 支持生成pdf、zip
- [x] 支持发送到邮箱
- [x] 支持设置代理
- [x] 支持API调用 [API-README](API-README.md)
- [x] 支持登录、下载自己账号下已购买的付费资源

## 安装/升级步骤

自己找安装Python的教程（Python版本大于等于3.6）
```sh
# 检查python版本
python --version
```

### 方式一（推荐）

```sh
# 在线安装/升级（最新版本）
# 由于网络环境不好导致安装失败 可以搜索关键词尝试解决: github host 修改
python -m pip install -U git+https://github.com/lossme/ComicBook

# 查看帮助
python -m onepiece --help

# 安装指定版本
python -m pip install git+https://github.com/lossme/ComicBook@v0.3.20
```

### 方式二（源码安装）

```sh
# clone项目 或从这里下载最新的代码并解压 https://github.com/lossme/ComicBook/releases
git clone git@github.com:lossme/ComicBook.git
# 切换工作目录
cd ComicBook

# 安装
python setup.py clean --all install

# 查看帮助
python -m onepiece --help

# 更新并安装至最新版本
git pull && python setup.py clean --all install
```

如果在使用过程中，发现问题可以先更新代码再试下，说不定已经修复了。

Star防止失联，欢迎大家提建议和issue，本项目持续更新。

## 常规使用

```sh
# 注意参数里的 - 和 -- 的区别
# 从章节列表页面的URL 下载漫画的最新一集
python -m onepiece --url "http://ac.qq.com/Comic/ComicInfo/id/505430"

# 下载漫画 id=505430 最新一集 注意不同站点的漫画id区别
python -m onepiece -s qq -id=505430

# 下载所有章节
python -m onepiece -s qq -id 505430 --all

# 下载第800集
python -m onepiece -s qq -id 505430 -c 800

# 下载倒数第二集
python -m onepiece -s qq -id 505430 -c -2

# 下载1到5集,7集，9到10集
python -m onepiece -s qq -id 505430 -c 1-5,7,9-10

# 拼接成长图
python -m onepiece -s qq -id 505430 --single-image --quality 95 --max-height 20000

# 设置代理
python -m onepiece -s qq -id 505430 --proxy "socks5://127.0.0.1:1080"

# 自定义保存目录
python -m onepiece -s qq -id=505430 --output MyComicBook

# 将多话合并到单个文件夹和zip文件
python -m onepiece -s manhuagui -id 1128 -c 320-322 --merge --merge-zip

# 下载单行本
python -m onepiece -s manhuagui -id 1128 --ext-name 单行本 -c -1


# 通过指定的URL文件列表批量下载
# 文件示例: https://github.com/lossme/ComicBook/blob/master/test/test-url-file.txt
python -m onepiece --url-file test/test-url-file.txt

# 跟据名字搜索comicid
python -m onepiece -s qq --name 海贼

# 生成pdf文件
# 注意: 生成pdf文件需要额外安装依赖，需要先执行 python -m pip install img2pdf 或 python -m pip install reportlab
python -m onepiece -s qq -id 505430 --pdf

# 推送到邮箱
# 注意: 发送到邮箱需预先配置好信息 配样例参照 https://github.com/lossme/ComicBook/blob/master/config.ini.example
# 并根据实际情况调整，将配置文件保存为 config.ini
python -m onepiece -s qq -id 505430 --pdf --mail --config config.ini
```

从其它站点下载，注意不同站点的comicid区别
```sh
# 从哔哩哔哩漫画下载
python -m onepiece -s bilibili -id mc24742 -c 1
# 从有妖气漫画下载
python -m onepiece -s u17 -id 195 -c 1

# 其它已支持的站点 https://github.com/lossme/ComicBook/projects/1
# 从章节列表页面的URL下载
python -m onepiece --url "https://manga.bilibili.com/detail/mc28603" -c 1
```

### 关于登录

1. [安装EditThisCookie插件](https://chrome.google.com/webstore/detail/editthiscookie/fngmhnnpilhplaeedifhccceomclgfbg)
2. 在浏览器上登录某个站点，然后通过插件导出某个站点的cookies，并保存到本地文件 如`qq.json`
```sh
python -m onepiece -s qq -id=505430 -c -1 --cookies-path="qq.json"
```

### 高级批量下载

```sh
# 有些站点不一定支持，其它的通用参数也适用，可自行组合
# 下载最近更新页面的1到10页 所有漫画的最新一集
python -m onepiece -s nvshens --latest-all --latest-page 1-10

# 展示支持的标签
python -m onepiece -s nvshens --show-tags

# 下载标签搜索结果页面的1到10页 所有漫画的全集
python -m onepiece -s nvshens --tag-all --tag 女神 --tag-page 1-10 --all

# 下载搜索结果的所有漫画的全集
python -m onepiece -s nhentai --search-all --search-name 汉化 --search-page 1 --all
```

### 高级配置

#### 通过环境变量配置默认参数
```sh
# 可以将以下命令添加到 ~/.bashrc 或 ~/.zshrc 文件末尾
# 配置默认的配置文件路径 配置文件示例 https://github.com/lossme/ComicBook/blob/master/config.ini.example
export ONEPIECE_CONFIG_FILE="/home/key/MyConfig/config.ini"
```

### 一个意义不明的群
https://t.me/onecomicbook


CHANGELOG.md
**unreleased**
**v0.3.23**
- toomics漫画下载优化
- fix 命令行下proxy参数不生效

**v0.3.22**
- 添加漫画源: 拷贝漫画 Toomics webtoons
- fix 动漫之家图片列表切割问题、url编码问题 #57

**v0.3.21**
- fix 合并章节排序错乱问题
- 根据搜索结果批量下载

**v0.3.20**
- 去除dukpy依赖
- 处理window平台下execjs编码错误的问题#52
- 配置文件编码指定为UTF-8

**v0.3.19**
- 添加MANIFEST.in文件

**v0.3.18**
- 添加漫画源: 禁漫之家 污污漫画 TwHentai
- 优化漫画柜单话、单行本下载问题

**v0.3.17**
- fix 文件夹名字带特殊字符报错问题
- 图集等单话的站点，取消章节名字

**v0.3.16**
- 增加超时配置、下载延时配置
- fix: dm5搜索结果遗漏

**v0.3.15**
- 优化配置文件的读取

**v0.3.14**
- 添加漫画源：二人次元动漫
- 支持通过url下载
- 支持多章节合并功能
- 新增获取所有站点的配置信息接口
- 新增通过URL解析出site和comicid接口
- 其它一些问题修复和优化

**v0.3.13**
- 添加漫画源 177漫画、18H漫画区
- 添加图集：Nsfwpicx、宅男女神、秀人网
- 根据最近更新页、标签页批量下载
- 生成长图优化等

**v0.3.12**
- 添加漫画源：土豪漫画网、漫画160、古风漫画网、36漫画网
- 支持给每个站点设置单独的默认代理
- 一些问题修复

**v0.3.11**
- 添加漫画源：COCO漫画、DM5、漫画DB
- 支持通过环境变量设置默认配置
- 支持当单行本、番外篇等扩展漫画下载

**v0.3.10**
- 优化安装方式，提供`setup.py`安装脚本
- webp图片校验优化
- 自动登录逻辑优化，降低各站点登录适配门槛
- api接口支持登录校验

**v0.3.9**
- 添加漫画源：acg456、mh1234、77mh、dmzj
- 支持对18comic的图片重组
- 支持生成zip文件
- 兼容不同版本的pdf生成工具

**v0.3.8**
- 支持通过API管理cookies
- 支持通过API配置代理配置
- fix 腾讯漫画标签获取问题
- 增强标签搜索功能
- 支持生成长图

**v0.3.7**
**v0.3.6**
- fix manhuagui标签搜索功能、封面图获取不到的问题
- 更换效率更高的pdf生成工具
- 添加manhuatai源

**v0.3.5**
- Added: 支持异步下载任务
- Added: 任务列表查询

**v0.3.4**
- Added: 支持聚合搜索接口
- Added: 支持标签过滤
- Improve: 接口返回值增加site、source_name字段
- Improve: 支持按站点配置代理
- Improve: 接入bumpveriosn版本管理
- Deprecate: 弃用图片下载缓存


onepiece/cli.py
import argparse
import os
import logging
import time

from .comicbook import ComicBook
from .crawlerbase import CrawlerBase
from .utils import (
    parser_chapter_str,
    ensure_file_dir_exists,
    merge_books,
    merge_zip_books
)
from .session import SessionMgr
from .worker import WorkerPoolMgr
from .utils.mail import Mail
from . import VERSION
from .config import CrawlerConfig

logger = logging.getLogger(__name__)
HERE = os.path.abspath(os.path.dirname(__file__))


def parse_args():
    """
    根据腾讯漫画id下载图片,默认下载海贼王最新一集。

    下载海贼王最新一集:
    python onepiece.py

    下载漫画 id=505430 最新一集:
    python onepiece.py -id 505430

    下载漫画 id=505430 所有章节:
    python onepiece.py -id 505430 -m all

    下载漫画 id=505430 第800集:
    python onepiece.py -id 505430 -c 800

    下载漫画 id=505430 倒数第二集:
    python onepiece.py -id 505430 -c -2

    下载漫画 id=505430 1到5集,7集，9到10集:
    python onepiece.py -id 505430 -i 1-5,7,9-10
    """

    parser = argparse.ArgumentParser(prog="onepiece")

    parser.add_argument('-id', '--comicid', type=str,
                        help="漫画id，如海贼王: 505430 (http://ac.qq.com/Comic/ComicInfo/id/505430)")
    parser.add_argument('--url', type=str,
                        help="漫画id，如海贼王: http://ac.qq.com/Comic/ComicInfo/id/505430")
    parser.add_argument('--url-file', type=str, help="漫画URL列表")

    parser.add_argument('--ext-name', type=str, help="如：番外篇、单行本等。具体得看站点支持哪些")

    parser.add_argument('--name', type=str, help="漫画名")

    parser.add_argument('-c', '--chapter', type=str, default="-1",
                        help="要下载的章节, 默认下载最新章节。如 -c 666 或者 -c 1-5,7,9-10")

    parser.add_argument('--worker', type=int, help="线程池数，默认开启4个线程池")

    parser.add_argument('--all', action='store_true',
                        help="是否下载该漫画的所有章节, 如 --all")

    parser.add_argument('--pdf', action='store_true',
                        help="是否生成pdf文件, 如 --pdf")
    parser.add_argument('--single-image', action='store_true',
                        help="是否拼接成一张图片, 如 --single-image")
    parser.add_argument('--quality', type=int, help="生成长图的图片质量，最高质量100")
    parser.add_argument('--max-height', type=int, help="长图最大高度，最大高度65500")

    parser.add_argument('--login', action='store_true',
                        help="是否登录账号，如 --login")

    parser.add_argument('--mail', action='store_true',
                        help="是否发送pdf文件到邮箱, 如 --mail。需要预先配置邮件信息。\
                        可以参照config.ini.example文件，创建并修改config.ini文件")

    parser.add_argument('--receivers', type=str, help="邮件接收列表，多个以逗号隔开")
    parser.add_argument('--zip', action='store_true',
                        help="打包生成zip文件")

    parser.add_argument('--config', help="配置文件路径")

    parser.add_argument('-o', '--output', type=str,
                        help="文件保存路径，默认保存在当前路径下的download文件夹")

    s = ' '.join(['%s(%s)' % (crawler.SITE, crawler.SOURCE_NAME) for crawler in ComicBook.CRAWLER_CLS_MAP.values()])
    site_help_msg = "数据源网站：支持 %s" % s

    parser.add_argument('-s', '--site', type=str, choices=ComicBook.CRAWLER_CLS_MAP.keys(),
                        help=site_help_msg)

    parser.add_argument('--verify', action='store_true',
                        help="verify")

    parser.add_argument('--driver-path', type=str, help="selenium driver")

    parser.add_argument('--driver-type', type=str,
                        choices=CrawlerBase.SUPPORT_DRIVER_TYPE,
                        help="支持的浏览器: {}.".format(
                            ",".join(sorted(CrawlerBase.SUPPORT_DRIVER_TYPE)))
                        )

    parser.add_argument('--cookies-path', type=str, help="读取或保存上次使用的cookies路径")

    parser.add_argument('--latest-all', action='store_true', help="下载最近更新里的所有漫画")
    parser.add_argument('--latest-page', type=str, help="最近更新的页数，如1-10，默认第1页")

    parser.add_argument('--show-tags', action='store_true', help="展示当前支持的标签")
    parser.add_argument('--tag-all', action='store_true', help="下载标签里的所有漫画")
    parser.add_argument('--tag', type=str, help="标签id")
    parser.add_argument('--tag-page', type=str, help="标签页数，如1-10，默认第1页")

    parser.add_argument('--search-all', action='store_true', help="下载搜索结果的所有漫画")
    parser.add_argument('--search-page', type=str, help="搜索结果页数，如1-10，默认第1页")
    parser.add_argument('--search-name', type=str, help="搜索的名字")

    parser.add_argument('--proxy', type=str,
                        help='设置代理，如 --proxy "socks5://user:pass@host:port"')

    parser.add_argument('--node-modules', type=str,
                        help="node_modules 模块目录")
    parser.add_argument('--merge', action='store_true', help="将多话合并成一个文件夹")
    parser.add_argument('--merge-zip', action='store_true', help="将多话合并成一个压缩包")
    parser.add_argument('--image-timeout', type=int, help="图片下载超时时间")
    parser.add_argument('--crawler-timeout', type=int, help="站点访问超时时间")
    parser.add_argument('--crawler-delay', type=int, help="每个章节下载时间间隔")
    parser.add_argument('--migrate', action='store_true', help="文件夹重命名(旧版本迁移用)")

    parser.add_argument('-V', '--version', action='version', version=VERSION)
    parser.add_argument('--debug', action='store_true', help="debug")

    args = parser.parse_args()
    return args


def init_logger(debug=False):
    level = logging.DEBUG if debug else logging.INFO
    logger = logging.getLogger()
    handler = logging.StreamHandler()
    formatter = logging.Formatter(
        "%(asctime)s %(name)s %(lineno)s [%(levelname)s] %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S'
    )
    handler.setFormatter(formatter)
    logger.addHandler(handler)
    logger.setLevel(level)
    return logger


def download_main(comicbook, output_dir, ext_name=None, chapters=None,
                  is_download_all=None, is_gen_pdf=None, is_gen_zip=None,
                  is_single_image=None, quality=None, max_height=None, mail=None,
                  receivers=None, is_send_mail=None, merge=None, merge_zip=None,
                  crawler_delay=None):
    is_gen_pdf = is_gen_pdf or mail
    chapter_str = chapters or '-1'
    chapter_numbers = parser_chapter_str(chapter_str=chapter_str,
                                         last_chapter_number=comicbook.get_last_chapter_number(ext_name),
                                         is_all=is_download_all)
    chapter_dirs = []
    for chapter_number in chapter_numbers:
        try:
            chapter = comicbook.Chapter(chapter_number, ext_name=ext_name)
            logger.info("正在下载 【{}】 {} 【{}】".format(
                comicbook.name, chapter.chapter_number, chapter.title or chapter.chapter_number))

            chapter_dir = chapter.save(output_dir=output_dir)
            chapter_dirs.append(chapter_dir)
            logger.info("下载成功 %s", chapter_dir)
            if is_single_image:
                img_path = chapter.save_as_single_image(output_dir=output_dir, quality=quality, max_height=max_height)
                logger.info("生成长图 %s", img_path)
            if is_gen_pdf:
                pdf_path = chapter.save_as_pdf(output_dir=output_dir)
                logger.info("生成pdf文件 %s", pdf_path)

            if is_send_mail:
                mail.send(subject=os.path.basename(pdf_path),
                          content=None,
                          file_list=[pdf_path, ],
                          receivers=receivers)
            if is_gen_zip:
                zip_file_path = chapter.save_as_zip(output_dir=output_dir)
                logger.info("生成zip文件 %s", zip_file_path)
        except Exception:
            logger.exception('download comicbook error. site=%s comicid=%s chapter_number=%s',
                             comicbook.crawler.SITE, comicbook.crawler.comicid, chapter_number)
        if crawler_delay:
            logger.info("crawler delay. sleep %ss", crawler_delay)
            time.sleep(crawler_delay)

    start = chapter_numbers[0]
    end = chapter_numbers[-1]
    if merge:
        merge_dir = comicbook.get_merge_dir(output_dir=output_dir, start=start, end=end, ext_name=ext_name)
        ensure_file_dir_exists(dirpath=merge_dir)
        merge_books(chapter_dirs=chapter_dirs, output_dir=merge_dir)
        logger.info("合并成单文件夹 %s", merge_dir)

    if merge_zip:
        merge_zip_path = comicbook.get_merge_zip_path(output_dir=output_dir, start=start, end=end, ext_name=ext_name)
        ensure_file_dir_exists(filepath=merge_zip_path)
        merge_zip_books(chapter_dirs=chapter_dirs, target_path=merge_zip_path)
        logger.info("合并成单个zip文件 %s", merge_zip_path)


def download_search_result(result, **kwargs):
    for citem in result:
        comicbook = ComicBook(site=citem.site, comicid=citem.comicid)
        comicbook.start_crawler()
        echo_comicbook_desc(
            comicbook=comicbook,
            ext_name=kwargs.get('ext_name')
        )
        download_main(comicbook=comicbook, **kwargs)


def download_latest_all(page_str, **kwargs):
    comicbook = kwargs.pop('comicbook')
    page_str = page_str or '1'
    for page in parser_chapter_str(page_str):
        try:
            result = comicbook.latest(page=page)
            logger.info('download_latest_all. page=%s result=%s', page, len(result))
        except Exception:
            logger.info('download_latest_all error. page=%s', page)
            continue
        download_search_result(result, **kwargs)


def download_tag_all(tag, page_str, **kwargs):
    comicbook = kwargs.pop('comicbook')
    page_str = page_str or '1'
    for page in parser_chapter_str(page_str):
        try:
            result = comicbook.get_tag_result(tag=tag, page=page)
            logger.info('download_tag_all. current page=%s result=%s', page, len(result))
        except Exception:
            logger.info('download_tag_all error. page=%s', page)
            continue
        download_search_result(result, **kwargs)


def download_search_all(name, page_str, **kwargs):
    comicbook = kwargs.pop('comicbook')
    page_str = page_str or '1'
    for page in parser_chapter_str(page_str):
        try:
            result = comicbook.search(name=name, page=page)
            logger.info('download_search_all. current page=%s result=%s', page, len(result))
        except Exception:
            logger.info('search error. current page=%s', page)
            continue
        download_search_result(result, **kwargs)


def download_url_list(config, url_file, **kwargs):
    comicbook = kwargs.pop('comicbook')
    with open(url_file) as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith('#'):
                continue
            url = line
            site = ComicBook.get_site_by_url(url=url)
            comicid = ComicBook.get_comicid_by_url(site=site, url=url)
            if not site or not comicid:
                logger.info('Unknown url. url=%s', url)
                continue
            try:
                comicbook = ComicBook(site=site, comicid=comicid)
                init_crawler(site=site, config=config)
                comicbook.start_crawler()
                echo_comicbook_desc(comicbook=comicbook, ext_name=kwargs.get('ext_name'))
                download_main(comicbook=comicbook, **kwargs)
            except Exception:
                logger.info('download comicbook by url error. url=%s', url)
                continue


def show_tags(comicbook):
    msg_list = []
    for t1 in comicbook.get_tags():
        category = t1['category']
        t1_msg_list = []
        for t2 in t1['tags']:
            msg = '{name}={tag}'.format(name=t2['name'], tag=t2['tag'])
            t1_msg_list.append(msg)
        msg = '{}:\n{}'.format(category, '\t'.join(t1_msg_list))
        msg_list.append(msg)
    logger.info('支持的标签\n%s', '\n'.join(msg_list))


def echo_comicbook_desc(comicbook, ext_name=None):
    last_chapter_number = comicbook.get_last_chapter_number(ext_name)
    last_chapter_title = comicbook.get_last_chapter_title(ext_name) or str(last_chapter_number)
    name = "{} {}".format(comicbook.name, ext_name) if ext_name else comicbook.name
    msg = ("{source_name} 【{name}】 更新至: {last_chapter_number:>03} "
           "【{last_chapter_title}】 数据来源: {source_url}").format(
        source_name=comicbook.source_name,
        name=name,
        last_chapter_number=last_chapter_number,
        last_chapter_title=last_chapter_title,
        source_url=comicbook.source_url)
    logger.info(msg)


def init_crawler(site, config):
    proxy = config.get_proxy(site=site)
    if proxy:
        logger.info('set proxy. %s', proxy)
        SessionMgr.set_proxy(site=site, proxy=proxy)
    if config.verify:
        logger.info('set verify. verify=True')
        SessionMgr.set_verify(site=site, verify=True)

    # 加载cookies
    cookies_path = config.get_cookies_path(site)
    if cookies_path and os.path.exists(cookies_path):
        SessionMgr.load_cookies(site=site, path=cookies_path)
        logger.info('load cookies success. %s', cookies_path)


def save_cookies(site, config):
    cookies_path = config.get_cookies_path(site)
    if cookies_path:
        ensure_file_dir_exists(filepath=cookies_path)
        SessionMgr.export_cookies(site=site, path=cookies_path)
        logger.info("cookies saved. path={}".format(cookies_path))


def migrate(comicbook_dir):
    for crawler_cls in ComicBook.CRAWLER_CLS_MAP.values():
        if crawler_cls.SINGLE_CHAPTER:
            dir1 = os.path.join(comicbook_dir, crawler_cls.SOURCE_NAME)
            for name1 in os.listdir(dir1):
                dir2 = os.path.join(dir1, name1)
                if not os.path.isdir(dir2):
                    continue
                for name2 in os.listdir(dir2):
                    dir3 = os.path.join(dir2, name2)
                    if not os.path.isdir(dir3):
                        continue
                    new_name = name2.split(' ')[0]
                    target = os.path.join(dir2, new_name)
                    os.rename(dir3, target)


def main():
    args = parse_args()
    init_logger(debug=args.debug)
    config = CrawlerConfig(args=args)
    if args.migrate:
        migrate(config.output)
        exit(0)

    if args.url:
        site = ComicBook.get_site_by_url(args.url)
        if not site:
            raise RuntimeError('Unknown url. url=%s' % args.url)
        comicid = ComicBook.get_comicid_by_url(site=site, url=args.url)
    else:
        site = args.site or 'qq'
        comicid = args.comicid

    WorkerPoolMgr.set_worker(worker=config.worker)
    CrawlerBase.DRIVER_PATH = config.driver_path
    logger.debug('set DRIVER_PATH. DRIVER_PATH=%s', config.driver_path)
    CrawlerBase.DRIVER_TYPE = config.driver_type
    logger.debug('set DRIVER_TYPE. DRIVER_TYPE=%s', config.driver_type)
    CrawlerBase.NODE_MODULES = config.node_modules
    logger.debug('set NODE_MODULES. NODE_MODULES=%s', config.node_modules)

    comicbook = ComicBook(site=site, comicid=comicid)
    comicbook.set_crawler_timeout(config.crawler_timeout)
    comicbook.set_image_timeout(config.image_timeout)

    if args.login:
        comicbook.crawler.login()
        save_cookies(site=site, config=config)
    if args.show_tags:
        init_crawler(site=site, config=config)
        show_tags(comicbook=comicbook)
        exit(0)

    if args.name:
        init_crawler(site=site, config=config)
        result = comicbook.search(name=args.name)
        msg_list = []
        for item in result:
            msg_list.append("comicid={}\t{}\tsource_url={}".format(
                item.comicid, item.name, item.source_url)
            )
        logger.info('\n%s', '\n'.join(msg_list))
        exit(0)

    if args.mail:
        is_send_mail = True
        mail = Mail.init(config.get_config_file())
    else:
        is_send_mail = False
        mail = None

    download_main_kwargs = dict(
        comicbook=comicbook,
        output_dir=config.output,
        chapters=args.chapter,
        is_download_all=args.all,
        is_gen_pdf=args.pdf,
        is_gen_zip=args.zip,
        is_single_image=args.single_image,
        quality=config.quality,
        max_height=config.max_height,
        mail=mail,
        ext_name=args.ext_name,
        is_send_mail=is_send_mail,
        receivers=args.receivers,
        merge=args.merge,
        merge_zip=args.merge_zip,
        crawler_delay=config.crawler_delay
    )

    if args.url_file:
        download_url_list(config=config, url_file=args.url_file, **download_main_kwargs)
    elif args.latest_all:
        init_crawler(site=site, config=config)
        download_latest_all(page_str=args.latest_page, **download_main_kwargs)
    elif args.tag_all:
        init_crawler(site=site, config=config)
        download_tag_all(tag=args.tag, page_str=args.tag_page, **download_main_kwargs)
    elif args.search_all:
        init_crawler(site=site, config=config)
        download_search_all(name=args.search_name, page_str=args.search_page, **download_main_kwargs)
    else:
        init_crawler(site=site, config=config)
        comicbook.start_crawler()
        echo_comicbook_desc(comicbook=comicbook, ext_name=args.ext_name)
        download_main(**download_main_kwargs)


if __name__ == '__main__':
    main()
